### PR TITLE
(BOLT-883) Override settings more directly

### PR DESCRIPTION
**Pull request by MikaelSmith**
Without this, we get errors like
```
The task failed with exit code 1:
/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:54:in `rescue in rescue in init_storage': Cannot remove /tmp/d20181003-855-156fn3k/state/state.yaml: No such file or directory @ unlink_internal - /tmp/d20181003-855-156fn3k/state/state.yaml (Puppet::Error)
	from /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:50:in `rescue in init_storage'
	from /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:45:in `init_storage'
	from /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/configurer.rb:205:in `run'
	from /tmp/c33af27e-1acd-495c-a495-304fd058ded3/apply_catalog.rb:81:in `block in <main>'
	from /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet/context.rb:65:in `override'
	from /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/puppet.rb:260:in `override'
	from /tmp/c33af27e-1acd-495c-a495-304fd058ded3/apply_catalog.rb:69:in `<main>'
```

Ensure directories are created from main settings, similar to how `puppet apply` performs setup.

### CODE REVIEW

**Comment by MikaelSmith** [comment]
Shouldn't that happen with https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/storage.rb#L45?

**Comment by nicklewis** [comment]
Ah yeah it should. That does make this a bit more mysterious.

File: libexec/apply_catalog.rb [hunk 1]
```diff
@@ -16,6 +16,7 @@
 puppet_root = Dir.mktmpdir
 moduledir = File.join(puppet_root, 'modules')
 Dir.mkdir(moduledir)
+Dir.mkdir(File.join(puppet_root, 'state'))
 cli = Puppet::Settings::REQUIRED_APP_SETTINGS.flat_map { |setting| ["--#{setting}", puppet_root] }
 cli << '--modulepath' << moduledir
 Puppet.initialize_settings(cli)
```
Comment (MikaelSmith) [inline libexec/apply_catalog.rb]: Should we include `Puppet.settings.use :main`?

**Comment by MikaelSmith** [comment]
Ok, I think this is better now.

**Comment by MikaelSmith** [comment]
Something deeper's going wrong here.

**Comment by MikaelSmith** [comment]
I've tested this with https://github.com/puppetlabs/bolt/pull/662, which gives me more confidence it'll work.

File: libexec/apply_catalog.rb [hunk 2]
```diff
@@ -13,34 +13,22 @@
 
 # Create temporary directories for all core Puppet settings so we don't clobber
 # existing state or read from puppet.conf
+Puppet.initialize_settings
 puppet_root = Dir.mktmpdir
+Puppet::Settings::REQUIRED_APP_SETTINGS.each do |setting|
+  Puppet[setting] = File.join(puppet_root, setting.to_s.chomp('dir'))
+end
```
Comment (adreyer) [inline libexec/apply_catalog.rb]: I think we should explicitly call puppet.settings.use( ??? ) to make it easier to understand and avoid regressions.

Not sure exactly where that should happen. before line 25 maybe?

**Pull request merged**